### ISSUE TITLE
Special characters in XML attributes not being escaped correctly

### ISSUE THREAD

**Issue by eug48** [nashwaan/xml-js#69]
Thanks very much for this very useful library. I've run into an issue since ampersands in XML attributes aren't being escaped by js2xml. This seems to have happened after the change described here: https://github.com/nashwaan/xml-js/issues/26#issuecomment-323543868 however the StackOverflow answer that is [referenced](https://stackoverflow.com/a/21574020/5619747) seems to be incorrect or just very badly structured? The spec extract that is quoted seems to contradict its recommendation and attributes with ampersands don't validate properly - e.g. try https://validator.w3.org/check with

```xml
<?xml version="1.0"?>
<elt attr="abc&def"></elt>
```

**Comment by nashwaan** [comment]
@eug48 , you are right. I got it wrong when I based my solution on that StackOverflow thread.
I wasn't aware about https://validator.w3.org/#validate_by_input+with_options online checking.

Now I can conclude correctly that the 5 characters: `&` `<` `>` `"` `'` should be replaced with `&amp;` `&lt;` `&gt;` `&quot;` `&#39;` in both text of tags AND attributes.

Will need to find time to fix this issue.

**Comment by Dico200** [comment]
I also ran into this issue.

**Comment by Shubu15** [comment]
@nashwaan When this issue will resolve?

**Comment by rdiazroman** [comment]
is there any update on this issue? thanks!

**Comment by dolanmiu** [comment]
Also experiencing this issue, raised here:

https://github.com/dolanmiu/docx/issues/2443

**Comment by SteffenLanger** [comment]
The repository for `xml-js` seems to be inactive. Since we want to keep using this otherwise great library, we built a workaround.

```TypeScript
export function js2XmlFixed(obj: Element | ElementCompact, options?: Options.JS2XML): string {
    return js2xml(obj, {
        attributeValueFn: (attributeValue) => escapeXml(attributeValue),
    });
}

/**
 * Ensure characters that would break the XML/HTML structure are encoded properly. This method should be applied to all values
 * inserted into XML/HTML.
 */
export function escapeXml(unsafeString: string): string {
    if (!unsafeString) {
        return unsafeString;
    }

    /**
     * Numbers etc. cannot be replaced. They can be returned without conversion.
     */
    if (typeof unsafeString !== 'string') {
        return unsafeString;
    }

    return unsafeString.replace(/[<>&'"]/g, function (c) {
        switch (c) {
            case '<':
                return '&lt;';
            case '>':
                return '&gt;';
            case '&':
                return '&amp;';
            case "'":
                return '&apos;';
            case '"':
                return '&quot;';
        }
    });
}
```

@nashwaan, in case you are still out here, maybe this is a good foundation for a library-internal fix.

**Comment by DecimalTurn** [comment]
Using `attributeValueFn` option to supply a custom function seems to be a good workaround, but the example above has an issue. Since xml-js does already escape double-quotes, you might have `&quot;` in your attributes, so you can't replace all `&` otherwise you might get `&amp;quot;`.

An alternative approach is to use a negative lookahead:

```ts
function escapeXml(v: string): string {
	return v
		.replace(/&(?!quot;)/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/'/g, "&apos;");
}
```
